[Spring Data JPA] 망고 미션 1단계 제출합니다.

llffmmgg opened the pull request:
## 안녕하세요, 콜리! 만나서 반가워요 ㅎㅎㅎ!
마지막 기간 동안 잘 부탁드립니다. 😄

## PR - STEP 4

### 구현 사항

- **JDBC -> JPA 전환**
    -  여기서는 findAll 관련 메서드들을 List로 받아주는 것이 너무 편리해서 `JpaRepository`를 선택했고, `JpaRepository`는 기본적인 CRUD 기능뿐만 아니라 페이징, 정렬 기능도 내장하고 있고 확장성이 좋다고 생각해서 선택했습니다 !
- **auth 패키지 관련 수정**
   - 기존에 AuthService에서 처리하던 JWT 관련 로직은 클라이언트와 밀접한 관련이 있다고 판단하여, AuthController로 이동했습니다.
- **soft delete(`Theme`, `ReservationTime`)로 제공된 코드 유지 이유** 
  - 테마와 시간은 해당 테마의 방이 다 차거나 해당 시간의 모든 방이 차면 사용자에게 없다고 표시되며 실제로 해당 개념은 사라지는 것이 아니기 때문
  - 예약(`Reservation`)과 멤버(`Member`)는 요구사항에 따라 달라질 것 같아 기존 구현 그대로 두었습니다 !
---

2주동안 잘 부탁드립니다 🙇 
리뷰에따라 잘 학습하고 개선해볼게요 !

감사합니다 ㅎㅎㅎ

llffmmgg committed:
refactor: jpa 관련 의존성  ㅜ가

llffmmgg committed:
refactor: jap 관련 설정 추가

llffmmgg committed:
refactor: jpa 사용을 위한 엔티티 클래스 설정 및 스키마 변경

llffmmgg committed:
refactor: ReservationRepository 생성 및 Dao 삭제

llffmmgg committed:
refactor: repository 사용

llffmmgg committed:
refactor: theme 엔티티 jpa 적용

llffmmgg committed:
refactor: member, jpa 적용

llffmmgg committed:
refactor: service에서 클라이언트 관련된 코드를 알고 있어 컨트롤러로 옮김

jwt를 사용하면서 loginCheck와 같은 api 호출 시 db 조회을 하고 있음
그럴거면 세션을 쓰는 게 낫다고 생각해서 호출하는 부분을 삭제

llffmmgg committed:
refactor: member를 조회를 id로 변경

name은 중복될 수 있음, auth 변경에 따른 테스트 코드 변경

llffmmgg committed:
refactor: 애플리케이션 코드 수정에 따른 테스트 코드 수정

llffmmgg committed:
refactor: 지연로딩 추가

llffmmgg committed:
refactor: 컬럼 조건 추가

llffmmgg committed:
refactor: theme, time 기존 구현대로 soft delete로 변경

coli-geonwoo commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -45,8 +56,9 @@ public ResponseEntity<Void> login(@RequestBody @Valid LoginRequest loginRequest)
 
-        LoginCheckResponse result = authService.checkAccessToken(accessToken);
+        LoginCheckResponse result = new LoginCheckResponse(name);
         return ResponseEntity.ok()
```
[제안 🔈 ]

다음과 같이 ResponseEntity.ok(body)로 body 값을 바로 담을 수 있습니다.

```suggestion
    return ResponseEntity.ok(result);
```

llffmmgg replied:
반영하겠습니다 ! ㅎㅎ

coli-geonwoo commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,12 +1,31 @@
+    @Column(nullable = false)
+    private String role;
+
+    public Member() {}
```
[질문 ✋ ]

- 왜 JPA Entity에서는 기본 생성자가 필요할까요?
- 접근제어자가 꼭 public 만 가능할까요?

ref) https://it-jin-developer.tistory.com/60#JPA%EC%99%80%20%EC%9C%A0%EC%82%AC%ED%95%98%EA%B2%8C%20%EA%B8%B0%EB%B3%B8%20%EC%83%9D%EC%84%B1%EC%9E%90%EB%A5%BC%20%EA%B0%95%EC%A0%9C%ED%95%98%EB%8A%94%20%EA%B2%BD%EC%9A%B0%EB%A5%BC%20%EC%82%B4%ED%8E%B4%EB%B3%B4%EC%9E%90.-1

llffmmgg replied:
좋은 레퍼런스 감사합니다 ! ㅎㅎ 

> 왜 JPA Entity에서는 기본 생성자가 필요할까요?

JPA에서 DB에서 조회한 데이터를 Entity 객체로 매핑할 때 리플랙션을 사용합니다. 

Reflection API는 구체적인 클래스 타입을 알지 못해도 이름만을 통하여 클래스에 대한 정보를 얻고 접근할 수 있도록 하는 API 입니다. 일반적으로는 객체에 데이터를 넣기 위해서는 new 생성자 또는 setter가 필요하고, private 인 필드는 데이터를 넣지 못할수도 있습니다. 

Reflection는 기본 생성자를 통해 객체를 생성하고 setter를 사용하지 않고도 데이터를 넣을 수 있고, private 메서드에 접근 할 수 있습니다. 

결국 JPA는 Reflection API를 사용하여 동적으로 객체를 생성하기 때문에 기본 생성자를 강제합니다 !  
(JPA는 어떤 타입으로 엔티티 객체를 생성할 지 모르기때문에 Reflection을 사용합니다 !)

> 접근제어자가 꼭 public 만 가능할까요?

public 또는 protected가 가능합니다. 

JPA는 즉시 로딩, 지연 로딩이 있습니다. 
지연 로딩을 할 때 연관된 엔티티에 프록시(가짜 객체)를 넣은 뒤 객체를 실제로 사용할 때 (`xxx.getName()`) 엔티티를 조회하는 방식입니다. 

이 경우에 프록시 객체는 진짜 엔티티의 상속을 받아 해당 기본 생성자를 사용하는 객체이기 때문에 부모의 기본 생성자가 private이면 프록시 객체가 생성이 되지 않습니다 ! 

> 어떤 접근 제어자 ? 

객체의 일관성을 유지하기 위해 setter를 지양하는 경우가 많고 무분별한 객체 생성을 막기 위해 기본 생성자의 접근 제어자를 최대한 범위를 줄이는 게 좋기 때문에 protected를 사용하는 것이 좋을 것 같습니다 !

coli-geonwoo commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,12 +1,31 @@
+    private String password;
+
+    @Column(nullable = false)
+    private String role;
```
[추가미션 🔑 ]

해당 Role을 Enum Type으로 바꿔주세요!

ex) Role.USER, Role.ADMIN

llffmmgg replied:
반영하겠습니다 ~!

coli-geonwoo commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,14 +1,33 @@
+    @Column(nullable = false)
     private LocalDate date;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
[질문 ✋ ]

- Fetch 란 무엇인가요?
- FetchType을 LAZY로 설정한 이유는 무엇인가요?
- FetchType EAGER와 LAZY는 어떤 차이가 있나요?

ref) https://ict-nroo.tistory.com/132

llffmmgg replied:
> Fetch 란 무엇인가요?

사전적 의미는 '~을 가지고 오다' 입니다. 
JPA에서는 Fetch는 엔티티의 필드에 DB에서 실제 값을 가져오는 것 이고 여러 가지 방법으로 가져올 수 있습니다 ! 

> FetchType EAGER와 LAZY는 어떤 차이가 있나요?

#### EAGER 

특정 엔티티를 조회할 때 연관된 모든 엔티티를 같이 가져옵니다. 
엔티티 간의 관계가 복잡해질수록 조인으로 인한 성능 저하가 있을 수 있습니다.   

#### LAZY

엔티티를 조회할 때 연관된 엔티티들을 프록시 객체로 가져오고, 실제 사용할 때 DB에서 조회합니다. 

> FetchType을 LAZY로 설정한 이유는 무엇인가요?

우선적으로는 LAZY로 설정하면 위에 적힌대로 해당 엔티티를 실제 사용할 때 DB에서 조회하기 때문에 테이블 관계가 복잡한 실무에서 대부분 사용한다고 알고 있어서 사용했어요 !

더 찾아봤을 때, ManyToOne의 fetchType 기본 값은 EAGER이고 OneToMany의 fetchType 기본 값은 LAZY 라는 것을 확인했어요. 

> 정확한 이유는 모르지만, ManyToOne이면 하나의 엔티티를 조회할 때 추가적으로 가져오는 데이터는 ManyToOne이 붙어있는 엔티티만큼이지만, 
OneToMany인 경우에는 하나의 테마에 1억개 예약이 있으면 EAGER인 경우 한번에 1억개를 더 조회하게 되고 사용하지 않는 다면 손해가 크다고 생각해서 그런 것 같아 저런 Default 값을 설정을 한 것 같기도 합니다..

불필요한 로딩을 피하는 관점에서는 LAZY가 좋고 이제 N+1 문제를 생각해보면 
EAGER를 사용해도 `ReservationRepository.findAll()` 을 할 때 실제 쿼리가 n+1 형태로 추가 쿼리가 나가는 것을 확인해서 결국 fetch join 처럼 해당 문제를 해결해야합니다. 
따라서 기본적으로 LAZY를 사용하고, 필요한 경우 fetch join, EntityGraph 등을 통해 명시적으로 최적화하는 것이 좋은 방법이라고 생각하게 되었습니다 !

coli-geonwoo replied:
[키워드 🔑  ]

망고의 답변에서 흥미로운 키워드들이 많네요! 다음과 같은 부분을 조금 더 고민해보죠

**- 1) LAZY로 설정한다고 N+1 문제가 해결될까요? N+1 문제란 무엇인가요?**
- ref)  https://incheol-jung.gitbook.io/docs/q-and-a/spring/n+1
->  ⚠️  배치 사이즈와 FetchMode에 대한 이야기는 지금은 학습하지 않으셔도 괜찮습니다.

**- 2) fetch join이 무엇인지 설명해주세요**

- ref) https://kihwan95.tistory.com/12

**- 3) fetch join과 entityGraph의 차이를 설명해주세요 (참고로 직접 설정해보시고 sql문 확인해보시는 걸 추천드립니다.)**

ref) https://velog.io/@xogml951/EntityGraph%EC%99%80-fetch-join-%EC%B0%A8%EC%9D%B4

llffmmgg replied:
> LAZY로 설정한다고 N+1 문제가 해결될까요? N+1 문제란 무엇인가요?

N+1 문제란 1번의 쿼리로 N개의 엔티티를 조회했을 때, 해당 엔티티가 연관된 엔티티를 조회하기 위해 추가적으로 N번의 쿼리가 실행되는 것 처럼 추가 쿼리가 발생하는 문제를 말합니다. 

보통 fetch join, entityGraph를 사용하여 데이터를 한번에 가져오는 방식으로 해결합니다.  

따라서 조회할 때 해당 엔티티를 같이 가져오지 않기때문에 발생하는 문제라 LAZY 설정은 N+1 발생 시점을 연관 데이터 사용 시점으로 옮기는 것이지 해결하는 것은 아닙니다 !

> fetch join이 무엇인지 설명해주세요

**fetch join**은 JPQL에서 제공하는 조인입니다. 

**FetchType.LAZY**일 때 기본 join을 사용하면 연관 엔티티를 즉시 조회하지 않고, 프록시 객체만 반환하며, 실제 데이터는 필요할 때 별도의 SQL 쿼리로 데이터를 가져옵니다. 

[장점]
- **fetch join**을 사용하면 쿼리 실행 시점에 연관된 엔티티를 한 번의 쿼리로 연관된 엔티티를 모두 가져올 수 있습니다 !
- 쿼리 최적화로 인한 성능 향상이 있을 수 있습니다.

[단점]
- fetch 대상에 별칭을 사용하지 못합니다. (하이버네이트에서는 별칭을 지원하지만 jpa 표준에서는 지원하지 않아 주의해야합니다)
- 둘 이상의 컬렉션을 fetch join 하면 두개의 컬렉션이 카테시안 곱이 만들어지므로 많은 중복 행이 생성될 수 있고 하이버네이트를 사용한다면 에러를 반환합니다.
- OneToMany 관계에서 연관된 엔티티를 가져올 때, 조인으로 인해 Many 쪽 엔티티 수만큼 One 쪽 엔티티가 중복해서 나타날 수 있습니다.

> fetch join과 entityGraph의 차이를 설명해주세요 (참고로 직접 설정해보시고 sql문 확인해보시는 걸 추천드립니다.)

**entityGraph** 는 jpa에서 제공하는 기능으로, 연관된 엔티티를 함께 조회하기 위해 사용하는 기능입니다. 

Reservation을 조회하면서 ReservationTime도 함께 가져오는 테스트를 해봤습니다. fetch join은 inner join을 사용하고, entityGraph는 left join을 사용하는 걸 확인했어요.

[fetch join]
연관된 엔티티가 존재하지 않는 경우, inner join의 특성상 주 엔티티가 결과에서 누락될 수 있습니다. 이를 해결하려면 LEFT JOIN FETCH를 명시적으로 사용해 outer join으로 설정할 수 있었습니다.

[entityGraph]
기본이 outer join(left)이어서 연관된 엔티티가 존재하지 않는 경우에도 주 엔티티들을 모두 잘 가져오는 것을 확인할 수 있었습니다.

[distinct]
중복 데이터를 처리하기 위해 JPA는 DISTINCT를 제공합니다. 
JPA의 DISTINCT는 SQL 쿼리에서 중복 행을 제거하는 기능과 함께, 객체 수준에서 중복 엔티티를 필터링하는 역할을 합니다.

블로그나 다양한 경로로 찾아봤을 때 fetch join이나 entityGraph를 사용하여 OneToMany 관계의 엔티티를 조회하는 경우 join으로 인한 중복 데이터 문제가 있어 DISTINCT를 사용하거나 Set 자료형을 사용하여 해결한다고 확인했습니다. 

하지만 테스트 해보았을 때, DISTINCT를 안붙여도 entityGraph, fetch join 모두 기본적으로 중복처리가 되는 것을 확인했습니다..

많은 삽질 후 **hibernate 6** 이후 부터 변경되었다는 것을 확인했습니다 !

--- 

최대한 검토했지만.. 학습하며 작성한 내용이라 잘못된게 있을수도 있습니다 😭 

좋은 리뷰 감사합니다 ~!

coli-geonwoo replied:
잘 학습해주셨네요 망고 👏 

사실 이슈를 제대로 겪지 않으면 체감되지 않는 내용이지만 그럼에도 테스트를 만들어 간접체험함으로서 학습에 도움이 되셨길 바랍니다. 특히 feth join의 경우, inner join이 발생하고 entityGraph의 경우 outer join 이라는 점을 잘 알아두시면 원하는 어떤 결과나 쿼리를 날리기 원하느냐에 따라 join 전략을 다르게 할 수 있겠지요.

조금 더 간단히 정리하면 다음과 같습니다.

- entityGraph는 런타임에 LAZY 전략을 EAGER로 바꾸는 것입니다. 이는 다른 말로 EAGER 전략은 outer left join으로 조인전략을 사용한다는 것을 알 수 있습니다.
- fetch join은 inner join을 통해 가져온 연관관계 테이블을 프록시가 아닌 실제 객체로 채워주는 것을 말합니다. 

저도 실무 경험이 없어 모르겠지만 제가 참여한 대부분의 프로젝트에서는 일단 fetch join을 고려해보는 것 같습니다. 개인적인 경험으로는 일단 LAZY를 통해 N+1 문제를 지연하되, 필요한 경우에 한하여 fetch join 쿼리를 만들어 사용하는 경우가 많았습니다. 따라서 지금으로서는 LAZY fetch란 무엇인지, 그리고 상황에 따라 fetch join을 적용하는 시각만 길러도 충분하다 생각합니다.

coli-geonwoo commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -51,7 +70,7 @@ public ReservationTime getTime() {
     }
 
     public String getTimeValue() {
```
[질문 ✋ ]

해당 메서드는 Dto를 의식한 코드인 것 같네요. 만약 dto가 String이 아닌 다른 값이라면 domain이 변경되겠네요.
dto의 변화가 도메인으로 전파되는 코드, 바람직할까요?

llffmmgg replied:
완전 동의합니다 !  
출력양식의 변화는 dto의 책임이니깐 도메인으로 넘어가면 안된다고 생각하고 바람직하지 않다고 생각해요 !

단순히 `String`을 formatted() 사용해서 hh:mm 양식을 맞추면서 디미터 법칙도 지키기 위해 내부에서 굳이 String을 반환했는 데, `LocalTime`에서 format()을 사용하면 되는군요.. 제대로 찾아보지 못했었네요 ㅠㅠ

coli-geonwoo commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface ReservationRepository extends JpaRepository<Reservation, Long> {
+
+    List<Reservation> findByDateAndTheme_Id(LocalDate date, Long theme);
```
파라미터에 null이 들어갈 수 없다면 long으로 표기해주는 것이 명세에 좋습니다. 또한, 파라미터명이 잘못되어 있는 것 같네요!

```suggestion
    List<Reservation> findByDateAndTheme_Id(LocalDate date, long themeId);
```

llffmmgg replied:
감사합니다 ! 
원시 타입을 쓰면 null이 들어갈 수 없다는 것을 코드로써 확실히 해주니 다른 사람이 볼 때도 좋은 것 같네요 ㅎㅎ

모두 반영했습니다  !

coli-geonwoo commented on src/main/java/roomescape/reservation/ReservationService.java:
```diff
@@ -2,38 +2,42 @@
     public List<ReservationResponse> findAll() {
-        return reservationDao.findAll().stream()
+        return reservationRepository.findAll().stream()
                 .map(it -> new ReservationResponse(it))
```
it은 무슨 뜻일까요? 조금 더 의미를 담아보죠

llffmmgg replied:
반영했습니다 !
it -> reservation

coli-geonwoo commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -1,49 +1,51 @@
                 .map(time -> new AvailableTime(
                         time.getId(),
-                        time.getValue().toString(),
+                        time.getTimeValue().toString(),
```
[의견 🔈 ]

depth가 너무 깊어 코드 읽기가 쉽지 않은 것 같아요. 몇가지 로직을 분리해주면 좋을 것 같아요,

llffmmgg replied:
동의합니다 ㅠㅠ

각 시간(ReservationTime)을 AvailableTime으로 바꾸는 역할을 toAvailableTime이 하고,
해당 시간이 예약되었는지 여부는 isTimeBooked에서 판단하도록 분리했습니다!

coli-geonwoo commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -1,49 +1,51 @@
-                        time.getValue().toString(),
+                        time.getTimeValue().toString(),
                         reservations.stream()
                                 .anyMatch(reservation -> reservation.getTime().getId().equals(time.getId()))
```
[퀴즈 ❓ ]
```suggestion
                                .anyMatch(reservation -> reservation.getTime().getId().equals(time.getId()))
```

이 로직을 실행할 때 추가 쿼리가 나갈까요? 나가지 않을까요?

ref) https://tecoble.techcourse.co.kr/post/2022-10-17-jpa-hibernate-proxy/

llffmmgg replied:
아하 ! id를 초기화할 때는 프록시가 초기화되지 않아서 추가 쿼리가 나가지 않는군요 👍 

초기화 과정에서 id를 자바 빈 규약에 맞는 (getxxx) 로 호출하면 getIdentifier가 호출하게 되서 가지고 있던 id를 제공하는 것으로 이해했습니다. (내부 구조는 더 살펴봐야될 것 같아요 ㅜㅜ)

위에서 EAGER, LAZY 관해 학습할 때 fetchType에 따라 어떤 쿼리가 발생하는 지 테스트 코드 통해 확인했었어요. 
그때, xxx.getId() (`xxx = y.getxxxx()`)를 하면 추가 쿼리가 발생하지 않아 getTimeValue()로 변경해서 실험했었는 데 이런 이유가 있었군요 

감사합니다 !!

coli-geonwoo replied:
현재 수준에서는 너무 깊이 이해하지 않아도 될 것 같습니다. 망고가 이해한대로 리뷰의 의도는
- 자바 빈 규약에 맞는 identifier 조회 (ex. getId)시 프록시가 초기화되지 않아 추가 쿼리가 날아가지 않는다
- identifier가 아닌 다른 필드값을 조회하려 시도할 때 추가쿼리가 날아간다.

정도만 캐치하셔도 100% 가져가신 거라 생각합니다 😄

coli-geonwoo commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -1,49 +1,51 @@
 
     public ReservationTime save(ReservationTime reservationTime) {
-        if (reservationTime.getId() == null || reservationTime.getValue() == null) {
+        if (reservationTime.getId() == null || reservationTime.getTimeValue() == null) {
```
[질문 ✋ ]
getId가 null일 경우 저장하지 못하네요. 기본키 생성전략인 Identity 전략이 무엇인지 설명해주세요!

llffmmgg replied:
앗 수정했습니다. 필요없는 부분이었네요. 

> 기본키 생성전략인 Identity 전략이 무엇인지 설명해주세요!

Identity 전략은 데이터베이스가 기본 키를 오토 인크리먼트로 자동 생성하는 JPA의 기본 키 생성 방식입니다 ! 

Identity 전략을 사용하면, 엔티티의 id를 null로 설정해서 저장 시 데이터베이스가 값을 채웁니다.

만약 id 값을 미리 설정해두면, 기본 키 중복 오류나 저장 실패가 발생할 수 있습니다!

---

반영했습니다 ~!

coli-geonwoo commented on src/main/resources/application.properties:
```diff
@@ -3,9 +3,9 @@ spring.h2.console.enabled=true
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.ddl-auto=create-drop
+spring.jpa.defer-datasource-initialization=true
```
EOF 신경써주시면 좋을 것 같습니다~

llffmmgg replied:
앗 감사합니다 반영했어요 신경쓰겠습니다 ~!!

coli-geonwoo commented on src/main/resources/schema.sql:
```diff
@@ -1,55 +1,16 @@
+INSERT INTO reservation (name, date, reservation_time_id, theme_id)
 VALUES ('어드민', '2024-03-01', 1, 1),
        ('어드민', '2024-03-01', 2, 2),
        ('어드민', '2024-03-01', 3, 3);
```
마찬가지로 EOF 신경써주세요!

coli-geonwoo reviewed (changes requested):
안녕하세요 망고! 2주간 잘 부탁드립니다 🙇 

몇가지 리뷰 남겼으니 확인 부탁드려요!

llffmmgg committed:
refactor: 기본 생성자의 접근제어자: protected

llffmmgg committed:
refactor: enum 적용

EnumType.String을 선택
EnumType.ORDINAL를 사용하면 순서를 (1,2..) 저장하는 데 Enum에 새로운 값이 추가되면 그에 따라 순서를 맞춰줘야해서 String을 사용

llffmmgg committed:
refactor: fetchType=LAZY로 유지하며 N+1 해결

llffmmgg committed:
refactor: 불필요하게 String을 반환하여 dto의 변화가 도메인으로 전파되는 것을 필드를 직접 반환하여 해결

llffmmgg committed:
refactor: 파라미터에 null이 들어갈 수 없는 곳 원시타입으로 표기

llffmmgg committed:
refactor: depth가 깊은 코드 메서드 분리로 해결

llffmmgg committed:
chore: private 메서드 위치 사용 메서드 아래로 이도

llffmmgg committed:
refactor: ReservationTime에게 동일한 엔티티인지 물어보는 메서드 추가

llffmmgg committed:
refactor: 기본키 생성 전략:identity로 하면서 id의 null 값 체크하는 로직을 삭제

llffmmgg committed:
chore: 개행 및 불필요한 import 제거

llffmmgg reviewed (commented):
안녕하세요 콜리 ! 
좋은 리뷰 감사합니다 ㅠㅠ 🙇 

제가 일정이 겹쳐 테스트 코드를 많이 작성하지 못해 기능 확인에 어려움이 있을 것 같아 죄송합니다.. 
리뷰 진행하면서 계속 최대한 테스트 코드를 추가해나갈게요 !

다시한번 좋은 리뷰와 레퍼런스 제공해주셔서 감사합니다 ~!

coli-geonwoo commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -23,16 +26,24 @@ public class AuthController {
 
-    public AuthController(AuthService authService, CookieProvider cookieProvider, CookieResolver cookieResolver) {
+    public AuthController(AuthService authService, CookieProvider cookieProvider,
+                          CookieResolver cookieResolver,
```
[제안 🔈 ]

개행 기준 하나로 통일해주시면 좋을 것 같습니다 😄

llffmmgg replied:
- 구글 스타일 가이드에 따라 100개의 문자를 넘긴 경우 
- dot(.) 을 사용하는 경우 

개행을 진행했습니다 ~ !

coli-geonwoo commented on src/main/java/roomescape/reservation/ReservationRepository.java:
```diff
@@ -0,0 +1,14 @@
+    List<Reservation> findByDateAndTheme_Id(LocalDate date, long themeId);
+
+    @Query("select r from Reservation r join fetch r.reservationTime")
+    List<Reservation> findAllWithReservationTime();
```
[질문 ✋ ]

fetch join을 사용하셨네요?! 어떤 점을 고려하셨나요? 👍

llffmmgg replied:
reservation과 연관된 reservationTIme 엔티티 조회 시 발생하는 추가 쿼리 문제(n+1)를 해결하기 위해 

한번의 쿼리로 연관된 엔티티까지 조회하는 fetch join을 사용했습니다. 

ManyToOne 관계의 엔티티만 fetch join하고 있어서 중복 데이터 관련 문제는 생각하지 않았습니다 !

coli-geonwoo replied:
좋은 고려 사안이네요. 즉, 이 쿼리를 통해 애플리케이션에서 reservationTime의 id를 제외한 내부 값을 항상 같이 사용한다면 지금처럼 fetch join이 좋은 전략이 될거에요.

현재는 reservation response에서 LocalTime time을 반환하니 항상 같이 사용될 수 밖에 없네요.

다만 reservation response에서 time이 id를 반환하는 식으로 바뀌면 어떨까요?

```java
public record ReservationResponse(
     Long id,
     String name,
     String theme,
     LocalDate date,
     long timeId //예약 시간 엔티티의 아이디를 반환함
) {
}

```

이 경우에는 굳이 fetch join 이 필요하지 않을거에요. 프록시 객체의 getId를 해도 추가쿼리가 나가지 않기 때문에 꼭 실제 객체를 채워서 가져올 필요가 없게 되는 것이죠.

중요한 건 fetch join을 사용하냐, 아니냐가 아니라 **어떤 맥락에서 해당 쿼리를 사용하고, 그 맥락을 고려했을 때 fetch join의 사용이 합리적인가? 를 판단하는 시각**이라 생각합니다.

coli-geonwoo commented on src/main/java/roomescape/reservation/ReservationResponse.java:
```diff
@@ -12,6 +12,6 @@ public record ReservationResponse(Long id,
     public ReservationResponse(Reservation reservation) {
         this(reservation.getId(), reservation.getName(), reservation.getThemeValue(),
-                reservation.getDate(), reservation.getTimeValue().formatted(TIME_FORMATTER));
+                reservation.getDate(), reservation.getTimeValue().format(TIME_FORMATTER));
```
이제 LocalTime 형식에 대한 변화가 생겨도 dto의 변화가 도메인으로 전파되지 않겠네요 👍

coli-geonwoo commented on src/main/java/roomescape/reservationTime/ReservationTime.java:
```diff
@@ -1,30 +1,45 @@
+        this.timeValue = LocalTime.parse(timeValue);
+    }
 
+    public boolean isSame(ReservationTime reservationTime) {
```
[질문 ✋ ]

동등성과 동일성의 차이는 무엇인가요?
id를 기반으로 equals 비교를 한 이유가 궁금합니다!

llffmmgg replied:
#### 동등성 (Equality)
객체가 내용적으로 같은지를 비교하는 것 입니다. 
보통 `equals()` 메서드를 오버라이드해서 객체의 상태(필드 값등)가 같으면 동등하다고 봅니다. 
#### 동일성(identity)
두 객체가 메모리상에서 동일한 인스턴스인지 비교합니다. 즉 메모리 상에서 동일한 위치에 있는 지 비교하는 것이고 `==` 연산자를 사용해 비교합니다.

---

equals 비교가 사용된 코드를 타고 올라가다 보면,
```
List<Reservation> reservations = reservationRepository.findByDateAndTheme_Id(date, themeId);
List<ReservationTime> reservationTimes = reservationTimeRepository.findAll();
```
equals 비교는 위 코드에서 파생 된 ` reservationTime`과 `reservations.getReservatioonTime`을 비교하고 있습니다.  
둘 다 DB에서 조회한 객체이므로 식별자(id)만 같다면 같은 객체라고 간주해도 된다고 생각하여 equals(동등성) 비교를 했습니다.

결국 객체를 비교하는 것이니 동일성(==)이 더 정석일 수 있지만, db에서 가져온 데이터 비교라 id(식별자)만 비교해도 될 것 같아서.. 

콜리 의견이 궁금합니다 ! ㅎㅎ

coli-geonwoo replied:
이건 리뷰어들도 의견 차이가 있는 부분인데요. 우리는 domain 관점(행동하는 능동적 객체)으로 해당 reservation을 대할 건지, entity 관점(데이터베이스의 한 행)으로 reservation을 바라볼 건지 그 비중을 정해야 합니다.

만약 domain 관점이라면 reservation은 DB에 저장되는 맥락보다 도메인에서 어떤 역할과 책임, 그리고 무엇을 의미하는 객체인지를 더 바라보아야 겠지요. 그런 관점에서 id만 같으면 같은 객체라고 논리적 동등성을 판단할 수 있을까요? 아닐거에요. 아마 time에 대한 비교도 같이 이뤄줘야 할 겁니다.

두번째로 망고가 이야기한 `db에서 가져온 데이터 비교`처럼 엔티티 관점을 유지한다면 사실상 PK인 identifier만 비교해도 같은 값이 보장되겠죠. 좋아요. 그럼 id만 비교하면 됩니다.

그럼 다시 조금 더 상위적 관점에서 domain 관점으로 프로그래밍 하는 것이 바람직한가? db 관점으로 프로그래밍하는 것이 바람직한가? 를 바라봅시다.

우리가 JPA와 같은 ORM을 도입하는 이유는 무엇일까요? 아마 기존 jdbcTemplate이 SQL 중심의 데이터 중점적으로 프로그래밍을 하게 되고 객체지향적으로 바라보지 않기 때문에, domain의 객체지향 생태계와 패러다임이 불일치 되는 문제를 완화하기 위함일거에요. 즉 우리는 데이터 중심 프로그래밍보다 객체지향 패러다임을 지향하기 위해 ORM을 사용합니다.

즉, OOP를 지향하기 위해 ORM을 사용하는데 오히려 데이터 중심으로 논리적 동등성을 비교하는 것은 의도의 충돌이 있지 않나... 라고 저는 개인적으로 생각합니다.

coli-geonwoo commented on src/main/java/roomescape/theme/ThemeController.java:
```diff
@@ -14,26 +14,27 @@
 public class ThemeController {
 
-    private ThemeDao themeDao;
+    private final ThemeRepository themeRepository;
```
[질문 ✋ ]
테마에서는 서비스와 소통하는 것이 아니라 바로 repository와 소통하는 이유가 무엇일까요?

llffmmgg replied:
시간 상 기존 코드를 아직 리팩토링 하지 못했습니다 ㅠㅠ 
다음 미션 진행하며 리팩토링 해볼게요 ~!!

coli-geonwoo reviewed (changes requested):
망고! 잘 학습해주셨습니다 ㅎㅎ 

엄선한 6개의 질문 남겼으니 한번 정도만 더 티키타카 가보시죠!

coli-geonwoo commented on src/main/java/roomescape/reservationTime/ReservationTimeService.java:
```diff
@@ -1,49 +1,61 @@
+        return reservationTimeRepository.save(reservationTime);
     }
 
     public void deleteById(Long id) {
```
[고민할 거리 🤔 ]
한가지 케이스에 대해서만 더 고민해봅시다

- 예약 시간을 지운다면 -> 그 예약시간을 사용하고 있는 reservation들은 어떻게 될까요?
한번 테스트 코드를 만들어 어떤 결과가 나오는지 생각해봅시다 ㅎㅎ

llffmmgg replied:
> 예약 시간을 지운다면 -> 그 예약시간을 사용하고 있는 reservation들은 어떻게 될까요?
한번 테스트 코드를 만들어 어떤 결과가 나오는지 생각해봅시다 ㅎㅎ

```
@SQLDelete(sql = "UPDATE reservation_time SET deleted = true WHERE id = ?")
@Where(clause = "deleted = false")
```
reservationTime은 현재 위 애노테이션을 엔티티에 추가하여 soft delete를 하고 있습니다. 

테스트 해 본 결과 
```
Hibernate: 
    select
        r1_0.id,
        r1_0.date,
        r1_0.name,
        r2_0.id,
        r2_0.deleted,
        r2_0.time_value,
        r1_0.theme_id 
    from
        reservation r1_0 
    join
        reservation_time r2_0 
            on r2_0.id=r1_0.reservation_time_id 
            and (
                r2_0.deleted = false
            )
```
마지막에서 2번째 줄 처럼` r2_0.deleted = false` 가 ReservationTime 엔티티의 조회 쿼리에 추가적인 조건으로 자동 삽입되어 삭제 된 예약은 조회 시 사라지고 있습니다. 

예약 시간이 사라져도(다 차서 비활성화 되거나 더 이상 해당 시간에 받지 않기 위해 주인이 예약 시간 삭제를 누른 경우) 해당 시간을 사용하고 있는 예약은 보여야한다고 생각합니다. 또한 실제로 예약 시간이 삭제되는 것은 아니라 생각하기 때문에 soft delete를 유지하고 예약 조회 쿼리를 `join fetch`에서 `left join fetch`로 변경하여 예약 시간을 조회할 수 없더라고 예약은 보이게 변경했습니다. 

이렇게 하면 진짜 예약 시간을 삭제하는 경우와 위의 경우를 [삭제], 위의 경우를 [마감 또는 비활성화] 버튼으로 구분해서 
[삭제]를 눌렀을 때 해당 시간을 가지는 예약도 모두 삭제되는 api를 만들면 될 것 같아요 ! 

--- 

ps) 일단 join fetch -> left join fetch로 변경은 했는 데, [삭제] 관련 된 내용도 구현하는 게 좋을까요 ?

coli-geonwoo replied:
지금 맥락 정도로도 충분한 것 같아요. 더 깊게 들어가진 맙시다.

다만 soft delete가 아닌 hard delete 의 경우를 고려하면 좋을 것 같아요. hard delete의 경우에는 reservation time과 외래키 제약조건이 걸려있기 때문에 **실제로 삭제가 되지 않습니다.**

즉, reservationTime을 참조하는 다른 객체들을 모두 삭제해준 이후에야 reservationTime의 삭제가 가능합니다.

coli-geonwoo commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,26 +1,45 @@
+    private String password;
+
+    @Column(nullable = false)
+    @Enumerated(EnumType.STRING)
```
[질문 ✋ ]

계속 덧붙이게 되네요 ㅎㅎ EnumType.String 말고 한가지 설정이 더 있을 텐데, String을 선택한 이유는 무엇일까요?

llffmmgg replied:
앗 이 부분도 학습했었는데 적지 않았었네요 ㅜㅜ

`EnumType.ORDINAL`
Enum 값의 순서에 따라 DB에 저장합니다. 

이 프로젝트의 코드로 생각하면 **ADMIN**은 0, **USER**는 1로 저장됩니다.

`EnumType.STRING`
Enum 이름 그대로 문자열로 저장합니다. 

> 선택 이유
 
**EnumType.ORDINAL**는 enum 순서가 저장됩니다. 순서가 바뀌거나 enum에 값이 추가되면 기존에 저장된 값과 매핑이 어긋나 잘못된 데이터가 조회되거나 예기치 않은 동작이 발생할 수 있습니다. 
따라서 STRING 타입을 사용했어요 !

coli-geonwoo replied:
정당한 이유네요. 합리적으로 잘 설명했습니다 망고 🥭

coli-geonwoo commented on src/main/java/roomescape/reservation/Reservation.java:
```diff
@@ -1,17 +1,40 @@
+    @Column(nullable = false)
     private LocalDate date;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
[질문 ✋ ]

`연관관계의 주인`이라는 개념이 무언인지 학습해봅시다. 또, ManyToOne과 OneToMany 가 둘다 존재하는 양방향 연관관계 상황에서 연관관계의 주인은 누가 되는지 이야기해주세요!

https://velog.io/@conatuseus/%EC%97%B0%EA%B4%80%EA%B4%80%EA%B3%84-%EB%A7%A4%ED%95%91-%EA%B8%B0%EC%B4%88-2-%EC%96%91%EB%B0%A9%ED%96%A5-%EC%97%B0%EA%B4%80%EA%B4%80%EA%B3%84%EC%99%80-%EC%97%B0%EA%B4%80%EA%B4%80%EA%B3%84%EC%9D%98-%EC%A3%BC%EC%9D%B8

llffmmgg replied:
[연관관계의 주인]

데이터베이스는 연관 관계가 있는 2개의 테이블을 외래 키 하나로 양방향으로 조회할 수 있습니다. 
한 엔티티가 외래 키의 주인이 되어 관리(등록 수정 삭제)를 하는 것을 연관관계의 주인이라고 합니다!

즉 객체에 없는 양방향 연관관계를 외래 키를 관리하는 객체를 정해서 객체 지향 관점에서 유지하기 위한 전략입니다. 

> ManyToOne과 OneToMany 가 둘다 존재하는 양방향 연관관계 상황에서 연관관계의 주인은 누가 되는지 이야기해주세요!

JPA에서 양방향 연관관계의 주인은 외래 키를 필드로 가지는 엔티티입니다. ManyToOne을 가진 엔티티가 외래 키를 포함하므로 연관관계의 주인이 되며, 등록, 수정, 삭제 등 외래 키를 관리합니다. 
반대로 OneToMany 쪽은 주인이 아니며, mappedBy 속성을 통해 주인 엔티티를 참조합니다.

---

한 번에 양방향 관계를 설정하는 메서드를 연관관계 편의 메서드라 하고, 연관 관계를 변경하는 경우에는 기존의 연관 관계를 삭제하는 코드를 추가해야합니다 !

llffmmgg committed:
chore: 개행 기준에 맞추어 수정 (100)

llffmmgg committed:
study: forStudy 테이블 추가(reservation과 양방향 관계를 가진 테이블)
>>
>> 1:n 조회 시 발생하는 중복 문제를 확인하기 위해 추가해서 결과를 확인해봤는 데 중복이 없었다.
>> 다시 찾아보니 hibernate 6 위 버전부터는 fetch join 시 발생하는 중복 문제를 hibernate에서 필터링 해준다는 것을 찾았다.. (늦게..)

llffmmgg committed:
study: 가져오는 collection이 비어있을 때 fetchjoin, entityGraph 동작 차이

llffmmgg committed:
study: fetch join 시 발생하는 중복 데이터 문제를 left join fetch로 해결

llffmmgg committed:
test: 예약 시간이 사라져도 예약은 유지된다

llffmmgg reviewed (commented):
안녕하세요 콜리 : ) 
생각할 부분이 많았던 좋은 리뷰 감사합니다 ㅎㅎ

코멘트 및 반영 했어요 ! 감사합니다 🙇 

---

ps) 학습하면서 미션과 상관없는 코드들이 생겼는 데 지우는 게 좋을까요 ?
-  ForStudy 클래스 : reservation  과 양방향 연관 관계를 가지고 OneToMany일 때 fetch join, entityGraph 사용한 메서드 결과를 위해 생성 
- JpaTest 내부 테스트 코드 : 쿼리문 확인하는 용도의 테스트들

llffmmgg committed:
refactor: jpaTest 수정
